src/tqec/plaquette/rpng_description.py

```diff
@@ -0,0 +1,245 @@
+    - a=z, p=z --> CZ   controlled by the ancilla and targeting the data
+    - a=y, p=z --> CY   controlled by the data    and targeting the ancilla
+    By default a=z.
+    """
+    r: Optional[ExtendedBasisEnum]
+    p: Optional[BasisEnum]
+    n: Optional[int]
+    g: Optional[ExtendedBasisEnum]
+    a: BasisEnum = BasisEnum.Z
+
+    @classmethod
+    def from_string(cls, rpng_string: str) -> 'RPNG':
+        """Initialize the RPNG object from a 4-character or 5-character string
+        
+        4-character -> simplified format -> RPNG
+        5-character -> extended format -> RAPNG
+
+        Note that the ancilla basis(a) is the second character, just before (p).
+
+        Note that any character different from a BasisEnum / ExtendedBasisEnum
+        value would result in the corresponding field being None.
+        """
+        a_str = 'z'
+        if len(rpng_string) == 4:
+            r_str, p_str, n_str, g_str = tuple(rpng_string)
+        elif len(rpng_string) == 5:
+            r_str, a_str, p_str, n_str, g_str = tuple(rpng_string)
+        else:
+            raise ValueError('The rpng string must be exactly 4- or 5-character long.')
+        # Convert the characters into the enum attributes (or raise error).
+        r = ExtendedBasisEnum(r_str) if r_str in ExtendedBasisEnum._value2member_map_ else None
+        a = BasisEnum(a_str) if a_str in BasisEnum._value2member_map_ else None
+        p = BasisEnum(p_str) if p_str in BasisEnum._value2member_map_ else None
+        n = int(n_str) if n_str.isdigit() else None
+        g = ExtendedBasisEnum(g_str) if g_str in ExtendedBasisEnum._value2member_map_ else None
+        # Raise error if anythiong but '-' was used to indicate None.
+        if not r and r_str != '-':
+            raise ValueError('Unacceptable character for the R field.')
+        if not p and p_str != '-':
+            raise ValueError('Unacceptable character for the P field.')
+        if p and not a:
+            raise ValueError('Unacceptable ancilla basis in the extended format (i.e. rapng).')
+        if not n and n_str != '-':
+            raise ValueError('Unacceptable character for the N field.')
+        if not g and g_str != '-':
+            raise ValueError('Unacceptable character for the G field.')
+        return cls(r, p, n, g, a)
+
+    
+    def get_r_op(self) -> Optional[str]:
+        """Get the reset operation or Hadamard"""
+        op = self.r
+        if op == None:
+            return None
+        elif op.value in BasisEnum._value2member_map_:
+            return f'R{op.value.upper()}'
+        else:
+            return f'{op.value.upper()}'
+
+    def get_g_op(self) -> Optional[str]:
+        """Get the measurement operation or Hadamard"""
+        op = self.g
+        if op == None:
+            return None
+        elif op.value in BasisEnum._value2member_map_:
+            return f'M{op.value.upper()}'
+        else:
+            return f'{op.value.upper()}'
```

Comment:
These two methods can very much be factored together. For example:
```suggestion
    def get_op(self, function: Literal["meas", "reset"]) -> Optional[str]:
        """Get the operation or Hadamard"""
        op = self.r if function == "reset" else self.g
        if op == None:
            return None
        elif op.value in BasisEnum:
            return ("R" if function == "reset" else "M") + op.value.upper()
        else:
            return f'{op.value.upper()}'
```

Reply:
Note that I did not test the above code, `Literal` is also not imported, so better apply it in your code editor if you think this is a good suggestion.

Reply:
You mentioned above if the Hadamard should also be part of the preparation. Perhaps we can wait until we decide on that point before merging the implementations (which may perhaps slightly differ). Also, I find it mnemonically easier to keep the descriptive names.

Reply:
> You mentioned above if the Hadamard should also be part of the preparation. Perhaps we can wait until we decide on that point before merging the implementations (which may perhaps slightly differ). 

On that, I am fine with both. Having a matching implementation here would be an argument for having the possibility to set `h` on both `R` and `G` in my opinion.

> Also, I find it mnemonically easier to keep the descriptive names.

Even if we restrict `h` to be only on `G`, the code should be factored. You can then re-use the factored code to implement `get_g_op` and `get_r_op`, just avoid as much as possible code duplication.